Lib/zipfile/__init__.py

```diff
@@ -563,6 +565,12 @@ def _decodeExtra(self, filename_crc):
                 except UnicodeDecodeError as e:
                     raise BadZipFile('Corrupt unicode path extra field (0x7075): invalid utf-8 bytes') from e
 
+            elif tp == 0x5455:
+                flags = extra[4]
+                if ln < 5:
```

Comment:
`ln` can be less than 5 and still be valid. The `flags` in the local header and central header should always be the same and may indicate that any combination of the 3 times are present (even though the central header can only store the mtime). So it is possible that the atime and ctime flags are set but not the mtime and the central header would only have the flags byte and no time data.

The info-zip description says the mtime 'must' be present in the central header if the mtime flag is set but I'd settle for a warning.

You could move this check above the `flags = extra[4]` line and change it to catch the potential (corrupt) case where `ln == 0`. If ln == 0 it is either going to raise an IndexError or return the first byte from the next extra field.